### PR TITLE
Fix shared-weight gradient double-counting in zero-bubble pipeline schedules

### LINKED ISSUE #180952: [pipelining] dW split path double-counts grad for shared weights
### 🐛 Describe the bug

`torch.distributed.pipelining._backward.stage_backward_input` / `stage_backward_weight` (the dW/dX-split path used by pipeline schedules that separate input-gradient and weight-gradient passes, e.g. ZeroBubble / 1F1B-V) produces **incorrect weight gradients** when a parameter is used at multiple sites in the forward graph AND those use sites lie on a parent-child path.

### Minimal reproducer

```python
import torch
import torch.nn as nn
from torch.distributed.pipelining._backward import (
    stage_backward_input,
    stage_backward_weight,
)


class Net(nn.Module):
    def __init__(self):
        super().__init__()
        self.w = nn.Parameter(
            torch.tensor([[0.4, -0.2], [0.7, 0.3]], dtype=torch.float32)
        )

    def forward(self, x):
        return (x @ self.w @ self.w).sum()   # w used at TWO matmul sites


# baseline
net_ref = Net()
x_ref = torch.tensor([[1.1, -0.7]], dtype=torch.float32, requires_grad=True)
out_ref = net_ref(x_ref)
dx_ref, dw_ref = torch.autograd.grad(out_ref, [x_ref, net_ref.w])

# split path
net_split = Net()
x_split = torch.tensor([[1.1, -0.7]], dtype=torch.float32, requires_grad=True)
out_split = net_split(x_split)
weights = [net_split.w]
dinputs, param_groups = stage_backward_input(
    stage_outputs_or_loss=[out_split],
    output_grads=None,
    input_values=[x_split],
    weights=iter(weights),
)
stage_backward_weight(weights=iter(weights), param_groups=param_groups)

print("dx_ref  :", dx_ref)
print("dx_split:", dinputs[0])
print("dw_ref  :", dw_ref)
print("dw_split:", net_split.w.grad)
```

### Observed output

```
dx_ref  : tensor([[-0.1200,  0.4400]])
dx_split: tensor([[-0.1200,  0.4400]])
dw_ref  : tensor([[ 0.1700,  1.0500],
                  [-0.5700, -1.1300]])
dw_split: tensor([[ 0.3900,  2.1500],
                  [-0.7100, -1.8300]])            <-- WRONG
```

`dx` matches. `dw` is wrong with `max |diff| = 1.10`.

### Expected behavior

`dw_split` should equal `dw_ref` (what `torch.autograd.grad` computes in one shot). The whole point of `stage_backward_input` + `stage_backward_weight` is to produce the same gradients as a regular backward, just split across two calls.

### Root cause analysis

At the forward level, `w` appears at two matmul sites, which in the autograd graph become two separate `MmBackward` nodes that share the same `AccumulateGrad` node for `w`:

- `MmBackward_1` — backward for `x @ w` (forward output: `h0`)
- `MmBackward_2` — backward for `h0 @ w` (forward output: `h1`)

`get_param_groups` correctly collects **both** matmul nodes into `param_group["intermediates"]` for `w`, and `stage_backward_input` registers a prehook at each to save the incoming gradient (`dL/dh0` and `dL/dh1` respectively).

The problem is in `stage_backward_weight` at the call:

```
https://github.com/pytorch/pytorch/blob/0cfb7e25c47f2d63a71c475b594b5252be5b0ddb/torch/distributed/pipelining/_backward.py#L294-L299
```

It invokes:

```python
torch.autograd.grad(
    outputs=valid_edges,             # GradientEdge(MmBackward_1), GradientEdge(MmBackward_2)
    inputs=(GradientEdge(AccGrad_w, 0),),
    grad_outputs=valid_grad_outputs, # saved dL/dh0, saved dL/dh1
)
```

**Because `MmBackward_2`'s forward output `h1` is a descendant of `MmBackward_1`'s forward output `h0`**, starting backprop from `MmBackward_2` with grad `dL/dh1` reaches `w` via two routes:

1. Directly through `MmBackward_2`'s weight input → contributes `h0ᵀ @ dL/dh1` to `dw`. ✓
2. Through `MmBackward_2`'s input `h0` → then through `MmBackward_1` → contributes `xᵀ @ (dL/dh1 @ wᵀ)` to `dw`.

But the saved `dL/dh0` (captured by the hook at `MmBackward_1` during `stage_backward_input`) already equals `dL/dh1 @ wᵀ` in this graph, so starting backprop from `MmBackward_1` with this saved grad adds **the same contribution** `xᵀ @ dL/dh0` again.

Net effect: the contribution from `MmBackward_1` is counted **twice**. Numerically:

```
matmul_1_contrib = xᵀ @ dL/dh0 = [[0.22, 1.10], [-0.14, -0.70]]
matmul_2_contrib = h0ᵀ @ dL/dh1 = [[-0.05, -0.05], [-0.43, -0.43]]

baseline        = matmul_1_contrib + matmul_2_contrib        = [[0.17, 1.05], [-0.57, -1.13]]
split (buggy)   = 2 * matmul_1_contrib + matmul_2_contrib    = [[0.39, 2.15], [-0.71, -1.83]]
```

Both match the printed output.

### Trigger conditions

Double-counting occurs iff **both** hold for some weight `w`:

1. `param_group["intermediates"]` for `w` contains two nodes `I_parent` and `I_child` where `I_child`'s forward output is a descendant of `I_parent`'s forward output; and
2. The backward path from `I_child` to `AccGrad_w` passes through `I_parent` (i.e. the same weight `w` participates in the subgraph between `I_parent` and `I_child`).

Condition (2) is what makes this a **shared-weight** issue rather than a general nested-intermediate issue.

### Related code

- `stage_backward_input`: https://github.com/pytorch/pytorch/blob/0cfb7e25c47f2d63a71c475b594b5252be5b0ddb/torch/distributed/pipelining/_backward.py#L177-L257
- `stage_backward_weight`: https://github.com/pytorch/pytorch/blob/0cfb7e25c47f2d63a71c475b594b5252be5b0ddb/torch/distributed/pipelining/_backward.py#L260-L311
- `get_param_groups`: https://github.com/pytorch/pytorch/blob/0cfb7e25c47f2d63a71c475b594b5252be5b0ddb/torch/distributed/pipelining/_backward.py#L89-L140
- The problematic call: https://github.com/pytorch/pytorch/blob/0cfb7e25c47f2d63a71c475b594b5252be5b0ddb/torch/distributed/pipelining/_backward.py#L294-L299

### Impact

Any pipeline schedule that uses the dW/dX-split backward (e.g. ZeroBubble variants and any custom schedule calling these helpers directly) will silently produce wrong weight gradients for models with tied/shared weights where the shared parameter is used at multiple forward sites on nested paths. Common real-world occurrences include:

- Weight-tying between encoder input embedding and decoder output projection when both appear on the same micro-batch compute path.
- Repeatedly-applied layers (weight sharing across iterations within one forward call).

### Versions

```
PyTorch version: 2.10.0+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
Python version: 3.10.12 (64-bit runtime)
Python platform: Linux-6.8.0-106-generic-x86_64-with-glibc2.35
Is CUDA available: False
Is XPU available: False

[pip3] numpy==2.2.6
[pip3] torch==2.10.0+cpu
[pip3] triton==3.6.0

Note: bug reproduces on CPU-only and is independent of device. Reproduced against
current main (SHA 0cfb7e25c47f2d63a71c475b594b5252be5b0ddb).
```

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @xmfan @weifengpy

### PR DESCRIPTION
**Pull request by arkadip-maitra**
Fixes #180952 

`torch.distributed.pipelining` scheduler `ScheduleInterleavedZeroBubble` split backward into separate input and weight phases via `stage_backward_input`  and `stage_backward_weight`. When stage's module has shared weights the double count happens.

Registered prehooks when param group has multiple intermediates during the weight phase that replace incoming gradients with the values saved during the input phase, preventing the double counting.

### CODE REVIEW

File: torch/distributed/pipelining/_backward.py [hunk 1]
```diff
@@ -281,6 +281,19 @@ def stage_backward_weight(
                     # pyrefly: ignore [bad-argument-type]
                     valid_grad_outputs.append(grad)
 
+        # prevent shared weights from double counting when
+        # intermediates are ancestor-descendant
+        handles = []
+        if len(param_group["intermediates"]) > 1:
+            for grads_tuple, intermediate in zip(
```
Comment (Skylion007) [inline torch/distributed/pipelining/_backward.py]: Zip should have strict=True kwarg.

**Review by Skylion007 — APPROVED** [review]

File: torch/distributed/pipelining/_backward.py (hunk 1 above)
Comment (Skylion007) [reply torch/distributed/pipelining/_backward.py]: Would this actually be cleaner as an contextlib exit stack? What if an error happens here, what will happen to the handles?

Comment (arkadip-maitra) [reply torch/distributed/pipelining/_backward.py]: yes it is cleaner. the handles would leak but that should be cleaned up by gc. 
the same pattern exists in `stage_backward_input` above, do you want me to change that in this pr too?

**Comment by arkadip-maitra** [comment]
@sanketpurandare can you take a look at this pr too pls, thanks!!

**Comment by sanketpurandare** [comment]
@claude review for bugs, silent incorrectness, clean design, test coverage and conciseness. suggest corrective yet minimal actions/code for this PR if required.

**Comment by sanketpurandare** [comment]
> **Claude finished @sanketpurandare's task in 2m 17s** —— [View job](https://github.com/pytorch/pytorch/actions/runs/26269165373)
> 
> ### Review: Fix shared-weight gradient double-counting in zero-bubble pipeline schedules
> * [x]  Read changed files and understand the fix
> * [x]  Check for bugs and silent incorrectness
> * [x]  Evaluate design cleanliness and conciseness
> * [x]  Assess test coverage
> * [x]  Post review findings
> 
> ### Summary
> This is a **clean, correct, and minimal fix**. The approach is sound and I don't see bugs or silent incorrectness. A few observations below.
> 
> ### Correctness Analysis
> The mechanism works as follows:
> 
> 1. During `stage_backward_input`, prehooks on each intermediate save the gradients flowing through that node (`param_group["grads"][i] = grad_inputs`).
> 2. During `stage_backward_weight`, `torch.autograd.grad` starts backward from each intermediate toward the weights. When weight W is used twice (e.g., `relu(x @ W) @ W`), the intermediates form an ancestor-descendant chain. Without the fix, backward from intermediate A would traverse through intermediate B, double-counting B's contribution to dW.
> 3. The new prehooks "clamp" each intermediate to its saved gradient, preventing ancestors from propagating through descendants.
> 
> The prehook lambda `lambda grad_outputs, sg=grads_tuple: sg` correctly:
> 
> * Captures `grads_tuple` via default argument (avoids late-binding closure bug)
> * Returns the saved tuple, which replaces incoming gradients during weight backward
> * Returning `None` (if a `grads_tuple` entry is None) means "don't modify" — correct fallback
> 
> The guard `if len(param_group["intermediates"]) > 1` ensures no overhead for the common case (single intermediate per param group).
> 
> ### Potential Issues
> **1. Set iteration order dependency (low risk, pre-existing)**
> 
> `param_group["intermediates"]` is a `set`. The zip at line 288-291 assumes iteration order matches how `param_group["grads"]` was populated (line 210-218). This works because the set isn't mutated between calls and CPython guarantees stable iteration within a process. This is the same assumption made by the pre-existing zip at line 275-276 — not a new risk.
> 
> **2. Handle leak on exception (low risk, pre-existing pattern)**
> 
> If `torch.autograd.grad` raises, the handles at line 326-327 won't be cleaned up. As discussed in the review thread, this matches the existing pattern in `stage_backward_input` (lines 208-255). An `ExitStack` would be cleaner but is out of scope for this bug fix.
> 
> ### Test Coverage
> The test (`test_stage_backward_weight_shared_weights`) is well-structured:
> 
> * Uses a minimal shared-weight module (`x @ W → relu → result @ W`)
> * Compares split backward against reference `loss.backward()`
> * Checks both input grads and weight grads
> 
> **Suggestion** (optional, not blocking): A second test case with three uses of the same weight or a mix of shared + non-shared weights would strengthen confidence, but the current test adequately exercises the fix.
> 
> ### Verdict
> **No blocking issues.** The fix is correct, minimal, and well-tested. Ship it.

@claude suggest minimal fixes for the potential issues pointed out.

**Review by sanketpurandare — CHANGES_REQUESTED** [review]
@arkadip-maitra  your PR is good, but would appreciate if you could address the minimal issue fixes and also add the try-cat in stage_backward_input for avoiding leakage and add meaningful error message.

**Comment by arkadip-maitra** [comment]
> @arkadip-maitra your PR is good, but would appreciate if you could address the minimal issue fixes and also add the try-cat in stage_backward_input for avoiding leakage and add meaningful error message.

thanks for the review, i made intermediates ordering explicit, added hook cleanup on exception and added context to `stage_backward_input` failures

**Review by sanketpurandare — APPROVED** [review]
